zerver/lib/actions.py

```diff
@@ -1564,14 +1564,22 @@ def send_stream_creation_event(stream: Stream, user_ids: List[int]) -> None:
 def create_stream_if_needed(realm: Realm,
                             stream_name: Text,
                             invite_only: bool=False,
+                            history_public_to_subscribers: Optional[bool]=None,
                             stream_description: Text="") -> Tuple[Stream, bool]:
+    if history_public_to_subscribers is None:
+        if invite_only:
+            history_public_to_subscribers = False
+        else:
+            history_public_to_subscribers = True
+
```

Comment:
The correct logic for this, at the current stage, is the below:

```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index 04b9ecf..9dd10b6 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -1567,11 +1567,21 @@ def create_stream_if_needed(realm: Realm,
                             invite_only: bool=False,
                             history_public_to_subscribers: Optional[bool]=None,
                             stream_description: Text="") -> Tuple[Stream, bool]:
-    if history_public_to_subscribers is None:
-        if invite_only:
-            history_public_to_subscribers = False
-        else:
-            history_public_to_subscribers = True
+    if invite_only:
+        if history_public_to_subscribers is None:
+            # TODO: Once we have a UI for this feature, we'll remove
+            # settings.PRIVATE_STREAM_HISTORY_FOR_SUBSCRIBERS and set
+            # this to be False here
+            history_public_to_subscribers = settings.PRIVATE_STREAM_HISTORY_FOR_SUBSCRIBERS
+    else:
+        # If we later decide to support public streams without
+        # history, we can remove this code path.
+        history_public_to_subscribers = True
+
+    if realm.is_zephyr_mirror_realm:
+        # In the Zephyr mirroring model, history is unconditionally
+        # not public to subscribers, even for public streams.
+        history_public_to_subscribers = False
 
     (stream, created) = Stream.objects.get_or_create(
         realm=realm,
```

Three key reasons:
(1) You misread the carefully written code in models.py around `is_public`, which checks is_zephyr_mirror_realm in addition to `invite_only`, and thus that logic was being ignored here (which would have created problems for any newly created streams).
(2) Until the frontend offers an option for the new property explicitly in the UI, any users who were previously using `PRIVATE_STREAM_HISTORY_FOR_SUBSCRIBERS` would have had that setting be broken for any new streams created after we merge this commit (but before we added the UI).  So fixing this is a prerequisite to be able to merge these commits now, rather than having to wait for the UI commits to all be ready.
(3) Your original implementation enabled the unintended feature of supporting making a public stream without public history.  We may perhaps want such a feature eventually, but there's no reason to offer it by accident.

I think that's the only real issue with this PR, so I went ahead and made those tweaks, added tests for the new code paths, and merged the commit so far with those changes.